flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/BipartiteGraph.java

```diff
@@ -0,0 +1,272 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.graph;
+
+import org.apache.flink.api.common.functions.FilterFunction;
+import org.apache.flink.api.common.functions.MapFunction;
+import org.apache.flink.api.java.DataSet;
+import org.apache.flink.api.java.ExecutionEnvironment;
+import org.apache.flink.api.java.functions.KeySelector;
+import org.apache.flink.api.java.tuple.Tuple2;
+
+/**
+ *
+ * Bipartite graph is a graph whose vertices can be divided into two disjoint sets: top vertices and bottom vertices.
+ * Edges can only exist between a pair of vertices from different vertices sets. E.g. there can be no vertices between
+ * a pair of top vertices.
+ *
+ * <p>Bipartite graphs are useful to represent graphs with two sets of objects, like researchers and their publications,
+ * where an edge represents that a particular publication was authored by a particular author.
+ *
+ * <p>Bipartite interface is different from {@link Graph} interface, so to apply algorithms that work on a regular graph
+ * a bipartite graph should be first converted into a {@link Graph} instance. This can be achieved by using
+ * {@link BipartiteGraph#simpleTopProjection()} or
+ * {@link BipartiteGraph#fullBottomProjection()} methods.
+ *
+ * @param <TK> the key type of the top vertices
+ * @param <BK> the key type of the bottom vertices
+ * @param <TV> the top vertices value type
+ * @param <BV> the bottom vertices value type
+ * @param <EV> the edge value type
+ */
+public class BipartiteGraph<TK, BK, TV, BV, EV> {
```

Comment:
Would the generic parameters be easier to read as `public class BipartiteGraph<KT, KB, VVT, VVB, EV> {`?